engine/src/main/java/org/pentaho/di/trans/steps/randomccnumber/RandomCreditCardNumberGenerator.java

```diff
@@ -0,0 +1,309 @@
+  private static final int[] AMEX_LENGTH_LIST = new int[] { 15 };
+  private static final int[] DISCOVER_LENGTH_LIST = new int[] { 16 };
+  private static final int[] DINERS_LENGTH_LIST = new int[] { 14 };
+  private static final int[] ENROUTE_LENGTH_LIST = new int[] { 15 };
+  private static final int[] JCB_15_LENGTH_LIST = new int[] { 15 };
+  private static final int[] JCB_16_LENGTH_LIST = new int[] { 16 };
+  private static final int[] VOYAGER_LENGTH_LIST = new int[] { 15 };
+  private static final int[] AIRPLUS_LENGTH_LIST = new int[] {};
+  private static final int[] BANKCARD_LENGTH_LIST = new int[] { 16 };
+  private static final int[] MAESTRO_LENGTH_LIST = new int[] { 16 };
+  private static final int[] SOLO_LENGTH_LIST = new int[] { 16, 18, 19 };
+  private static final int[] SWITCH_LENGTH_LIST = new int[] { 16, 18, 19 };
+  private static final int[] LASER_LENGTH_LIST = new int[] { 16, 17, 18, 19 };
+
+  public static int getCardType( String typeName ) {
+    if ( typeName == null ) {
+      return 0;
+    }
+
+    for ( int i = 0; i < cardTypes.length; i++ ) {
+      if ( cardTypes[i].equalsIgnoreCase( typeName ) ) {
+        return i;
+      }
+    }
+    return 0;
+  }
+
+  public static String getCardName( int id ) {
+    return ( id > -1 && id < cardTypes.length ? cardTypes[id] : null );
+  }
+
+  private static String strrev( String str ) {
+    if ( str == null ) {
+      return "";
+    }
+    String revstr = "";
+    for ( int i = str.length() - 1; i >= 0; i-- ) {
+      revstr += str.charAt( i );
+    }
+
+    return revstr;
+  }
+
+  /*
+   * 'prefix' is the start of the CC number as a string, any number of digits. 'length' is the length of the CC number
+   * to generate. Typically 13 or 16
+   */
+  private static String completed_number( String prefix, int length ) {
+
+    String ccnumber = prefix;
+
+    // generate digits
+
+    while ( ccnumber.length() < ( length - 1 ) ) {
+      ccnumber += new Double( Math.floor( Math.random() * 10 ) ).intValue();
+    }
+
+    // reverse number and convert to int
+
+    String reversedCCnumberString = strrev( ccnumber );
+
+    List<Integer> reversedCCnumberList = new Vector<Integer>();
+    for ( int i = 0; i < reversedCCnumberString.length(); i++ ) {
+      reversedCCnumberList.add( new Integer( String.valueOf( reversedCCnumberString.charAt( i ) ) ) );
+    }
+
+    // calculate sum
+
+    int sum = 0;
+    int pos = 0;
+
+    Integer[] reversedCCnumber = reversedCCnumberList.toArray( new Integer[reversedCCnumberList.size()] );
+    while ( pos < length - 1 ) {
+
+      int odd = reversedCCnumber[pos] * 2;
+      if ( odd > 9 ) {
+        odd -= 9;
+      }
+
+      sum += odd;
+
+      if ( pos != ( length - 2 ) ) {
+        sum += reversedCCnumber[pos + 1];
+      }
+      pos += 2;
+    }
+
+    // calculate check digit
+
+    int checkdigit = new Double( ( ( Math.floor( sum / 10 ) + 1 ) * 10 - sum ) % 10 ).intValue();
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The result of `Integer` multiplication is cast into a larger type. If you think the result of the multiplication could exceed the size of an `Integer`, consider casting the arguments of the multiplication to avoid arithmetic overflows. For more information, see [CWE-190](https://cwe.mitre.org/data/definitions/190.html).